### ISSUE TITLE
FileLoadException: Could not load file or assembly ... Microsoft.AspNet.Mvc.Razor, Version=6.0.0.0

### ISSUE THREAD

**Issue by korzh** [aspnet/Mvc#4322]
We've got a strange exception recently running our MVC 6 application on MS Azure (as web app).
It runs fine right after publishing for several hours and then suddenly come up with the following error:

```
An unhandled exception occurred while processing the request.

UnauthorizedAccessException: Access is denied. (Exception from HRESULT: 0x80070005 (E_ACCESSDENIED))
Unknown location

FileLoadException: Could not load file or assembly '51712 bytes loaded from Microsoft.AspNet.Mvc.Razor, Version=6.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60' or one of its dependencies. Access is denied.
System.Reflection.RuntimeAssembly.nLoadImage(Byte[] rawAssembly, Byte[] rawSymbolStore, Evidence evidence, StackCrawlMark& stackMark, Boolean fIntrospection, SecurityContextSource securityContextSource)
```

I don't have any idea why it happens. It disappears on app restart but after several hours arises again.
This web-site is in development and don't have any traffic now except few requests per hour from developers and testers.
The detailed stack trace is in attachment.
[Error.zip](https://github.com/aspnet/Mvc/files/183679/Error.zip)

**Comment by Eilon** [comment]
Putting the stack trace inline here:

```
System.Reflection.RuntimeAssembly.nLoadImage(Byte[] rawAssembly, Byte[] rawSymbolStore, Evidence evidence, StackCrawlMark& stackMark, Boolean fIntrospection, SecurityContextSource securityContextSource)
System.Reflection.Assembly.Load(Byte[] rawAssembly, Byte[] rawSymbolStore)
Microsoft.AspNet.Mvc.Razor.Compilation.RoslynCompilationService.LoadStream(MemoryStream ms, MemoryStream assemblySymbols)
Microsoft.AspNet.Mvc.Razor.Compilation.RoslynCompilationService.Compile(RelativeFileInfo fileInfo, String compilationContent)
Microsoft.AspNet.Mvc.Razor.Compilation.RazorCompilationService.Compile(RelativeFileInfo file)
Microsoft.AspNet.Mvc.Razor.Compilation.CompilerCache.CreateCacheEntry(String normalizedPath, Func`2 compile)
Microsoft.AspNet.Mvc.Razor.Compilation.CompilerCache.GetOrAdd(String relativePath, Func`2 compile)
Microsoft.AspNet.Mvc.Razor.VirtualPathRazorPageFactory.CreateInstance(String relativePath)
Microsoft.AspNet.Mvc.Razor.RazorViewEngine.LocatePageFromViewLocations(ActionContext context, String pageName, Boolean isPartial)
Microsoft.AspNet.Mvc.Razor.RazorViewEngine.GetRazorPageResult(ActionContext context, String pageName, Boolean isPartial)
Microsoft.AspNet.Mvc.Razor.RazorViewEngine.FindView(ActionContext context, String viewName)
Microsoft.AspNet.Mvc.ViewEngines.CompositeViewEngine.FindView(ActionContext context, String viewName, Boolean partial)
Microsoft.AspNet.Mvc.ViewEngines.CompositeViewEngine.FindView(ActionContext context, String viewName)
Microsoft.AspNet.Mvc.ViewFeatures.ViewResultExecutor.FindView(ActionContext actionContext, ViewResult viewResult)
...
```

**Comment by Eilon** [comment]
@pranavkm this is super weird - any idea why the Razor view engine would try to load an assembly named `51712 bytes loaded from Microsoft.AspNet.Mvc.Razor, Version=6.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60`? How could what looks like some sort of log message make its way into the assembly name we try to load?

**Comment by pranavkm** [comment]
We attempt to load any dependency that was specified in the `DependencyContext`. Maybe there's a bug in it where it writes the assembly name incorrectly? @pakrym have you seen this?

**Comment by pranavkm** [comment]
Nevermind - this looks like loading the compiled Razor view assembly is failing. @korzh are you using the dotnet cli (dotnet.exe) to run your application?

**Comment by korzh** [comment]
I didn't understand at first that that "51712 bytes loaded..." is actually an assembly name. We will try to log the view content on this exception to find out when and why it happens.

**Comment by korzh** [comment]
This app is running as web-app on MS Azure. 
Target DNX version: dnx-clr-win-x86.1.0.0-rx1-final

**Comment by korzh** [comment]
I think I know the reason of the problem (at least I guess so).
We assigned our own IFileProvider which takes content for some pages (not all) from DB instead of files.
Watch method of that file providers generated IChangeToken which always returned "true" for HasChanged property. We did it for reason to make Razor to reload (and so - to recompile) page content from DB on each request. 
Recently we have changed this logic, so now HasChanged == true only when the page content was changed in DB since the last request. The problem has disappeared. 
Obviously, current implementation of Razor engine does not "like" if you make it to recompile the pages on each request. 
The problem occurs after some number of such requests but I can't say the exact amount now. 
Usually if we setup to "ping" one of our pages once a minute (just sending a GET request to it), the exception appears after 3-4 hours after such "pinging" so it we can estimate this number to about 200 requests. However I guess it depends on server configuration, amount of RAM and so on.
